### PR TITLE
fix: Transaction.GetTransactionID() panic if TransactionID not set, T…

### PR DESCRIPTION
**Pull request by andrix10**
…ransactionIDs and NodeAccountIDs not locking during Transaction.FromBytes()

Signed-off-by: Andrei Kuzmiankov <andrei@launchbadge.com>

Closes: #462
Closes: #463 
Closes: #454

### CODE REVIEW

File: transaction.go [hunk 1]
```diff
@@ -458,30 +484,28 @@ func _TransactionFreezeWith(
...
-	if err != nil {
-		// This should be unreachable
-		// From the documentation this appears to only be possible if there are missing proto types
-		panic(err)
-	}
-
-	for range transaction.nodeAccountIDs.slice {
+	for _, nodeAccountID := range transaction.nodeAccountIDs.slice {
+		body.NodeAccountID = nodeAccountID.(AccountID)._ToProtobuf()
+		bodyBytes, err := protobuf.Marshal(body)
```
Comment (xin-hedera) [inline transaction.go:503]: is there a test case?

the license header should really go to its separate PR. it's difficult to review the changes with so much noise.

File: transaction.go [hunk 2]
```diff
@@ -773,7 +797,12 @@ func (this *Transaction) SetTransactionValidDuration(duration time.Duration) *Tr
 }
 
 func (this *Transaction) GetTransactionID() TransactionID {
-	return this.transactionIDs._GetCurrent().(TransactionID)
+	if this.transactionIDs._Length() > 0 {
+		t := this.transactionIDs._GetCurrent().(TransactionID)
+		return t
+	}
+
+	return TransactionID{}
```
Comment (xin-hedera) [inline transaction.go:805]: is there a test case to prevent such regression?

File: transaction.go [hunk 3]
```diff
@@ -458,30 +484,28 @@ func _TransactionFreezeWith(
...
 			for _, nodeAccountID := range client.network._GetNodeAccountIDsForExecute() {
 				transaction.nodeAccountIDs._Push(nodeAccountID)
 			}
-		}
 
-		if client.defaultRegenerateTransactionIDs != transaction.regenerateTransactionID {
-			transaction.regenerateTransactionID = client.defaultRegenerateTransactionIDs
+			if client.defaultRegenerateTransactionIDs != transaction.regenerateTransactionID {
+				transaction.regenerateTransactionID = client.defaultRegenerateTransactionIDs
+			}
```
[the remark this answers is not in the document]
Comment (andrix10) [reply transaction.go:495]: yes, as if a user sets node account id, the regeneration is turned off

File: transaction.go (hunk 1 above)
Comment (andrix10) [reply transaction.go:503]: Yes license part was my bad, it was a fast one so i thought no point of creating another pr. I added the multi_app_test and [these.](https://github.com/hashgraph/hedera-sdk-go/pull/464/commits/9cb9e978952731a9bc4bfa0c8338514a29858ce0l)

File: transaction.go (hunk 2 above)
Comment (andrix10) [reply transaction.go:805]: Just had it added to one of the new unit tests.

File: transaction.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (andrix10) [reply transaction.go:495]: Nvm, for some reason I was thinking that node id gets modified in execute for regenerate so i wanted to simplify it, but it doesn't. I will be changing it back.

File: lockable_slice.go [hunk 4]
```diff
@@ -78,6 +78,7 @@ func (this *_LockableSlice) _Advance() int { //nolint
 	index := this.index
 	if len(this.slice) != 0 {
 		this.index = (this.index + 1) % len(this.slice)
+		index = this.index
```
Comment (danielakhterov) [inline lockable_slice.go:101]: Not sure if this is what we want to do. You're free to deviate from what we have In JS, but initially what I had in mind for `Advance` was to increment index and return previous index, not return the new incremented index. This change feels like it should cause some tests to fail and a bit concerned if that wasn't the case.

File: multi_app_transfer_test.go [hunk 5]
```diff
@@ -0,0 +1,70 @@
...
+
+	switch t := tx.(type) {
+	case TransferTransaction:
+		signedTx = t
+	default:
+		panic("Did not receive `TransferTransaction` back from signed bytes")
+	}
+
+	for _, s := range signedTx.GetNodeAccountIDs() {
+		println(s.String())
```
Comment (danielakhterov) [inline multi_app_transfer_test.go:61]: Shouldn't have print statements :)

File: transaction.go [hunk 6]
```diff
@@ -165,7 +165,13 @@ func TransactionFromBytes(data []byte) (interface{}, error) { // nolint
 		}
 	}
 
-	tx.transactionIDs.locked = true
+	if tx.transactionIDs._Length() > 0 {
```
Comment (danielakhterov) [inline transaction.go:188]: Maybe we should create a method `LockableList._LockIfNotEmpty()` so you don't have to repeat yourself?

File: lockable_slice.go (hunk 4 above)
Comment (andrix10) [reply lockable_slice.go:101]: It seems nothing actually uses the returned int.
